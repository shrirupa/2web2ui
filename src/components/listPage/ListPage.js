/* eslint-max-lines: ["error": 180] */
import React from 'react';

import { Page } from '@sparkpost/matchbox';
import {
  Loading,
  ApiErrorBanner,
  TableCollection,
  DeleteModal
} from 'src/components';
import { capitalize } from 'src/helpers/string';

import Actions from './Actions';
import ListPagePropTypes from './ListPage.propTypes';

const DEFAULT_STATE = {
  itemToDelete: null
};

class ListPage extends React.Component {
  state = DEFAULT_STATE;

  componentDidMount() {
    this.props.loadItems();
  }

  handleCancel = () => {
    this.setState(DEFAULT_STATE);
  };

  handleDelete = () => {
    const { onDelete } = this.props;
    const { itemToDelete } = this.state;

    this.setState(DEFAULT_STATE, () => {
      onDelete(itemToDelete);
    });
  };

  showDeleteModal = (item) => {
    this.setState({ itemToDelete: item });
  };

  renderError() {
    const { loadItems, error, noun } = this.props;
    return (
      <ApiErrorBanner
        errorDetails={error.message}
        message={`Sorry, we seem to have had some trouble loading your ${noun.toLowerCase()}s .`}
        reload={loadItems}
      />
    );
  }

  formatRowWithActions = (row) => {
    const { formatRow } = this.props;
    const formatted = formatRow(row);
    const fields = formatted.fields || formatted;
    const actions = formatted.actions || {};
    return [
      ...fields,
      <Actions
        item={row}
        editRoute={actions.editRoute}
        deletable={actions.deletable}
        onDelete={this.showDeleteModal}
        customActions={actions.customActions}
      />
    ];
  };

  renderItems() {
    const {
      columns,
      items,
      filterBox,
      defaultSortColumn
    } = this.props;
    const filterBoxOptions = filterBox ? { show: true, ...filterBox } : { show: false };
    return (
      <TableCollection
        columns={columns}
        getRowData={this.formatRowWithActions}
        pagination={true}
        rows={items}
        filterBox={filterBoxOptions}
        defaultSortColumn={defaultSortColumn || columns[0]}
      />
    );
  }

  renderDeleteWarning() {
    const { itemToDelete } = this.state;
    const { renderDeleteWarning } = this.props;

    if (!itemToDelete) {
      return null;
    }

    if (!renderDeleteWarning) {
      return null;
    }

    return renderDeleteWarning(itemToDelete);
  }

  render() {
    const { itemToDelete } = this.state;
    const {
      noun,
      primaryActionTitle,
      onCreate,
      loading,
      error,
      banner,
      empty,
      additionalActions
    } = this.props;
    const capsNoun = capitalize(noun);
    const primaryAction = onCreate && {
      content: primaryActionTitle || `Create ${capsNoun}`,
      onClick: onCreate
    };

    if (loading) {
      return <Loading />;
    }

    const modalContent = this.renderDeleteWarning();

    return (
      <Page
        title={`${capsNoun}s`}
        primaryAction={primaryAction}
        secondaryActions={additionalActions}
        empty={empty}
      >
        {banner}
        {error ? this.renderError() : this.renderItems()}
        <DeleteModal
          open={Boolean(itemToDelete)}
          onCancel={this.handleCancel}
          onDelete={this.handleDelete}
          title={`Are you sure you want to delete this ${noun}?`}
          content={modalContent}
        />
      </Page>
    );
  }
}

ListPage.propTypes = ListPagePropTypes;
ListPage.defaultProps = {
  items: []
};

export default ListPage;
