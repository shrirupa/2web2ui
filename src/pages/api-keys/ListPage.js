import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Page } from '@sparkpost/matchbox';
import { Setup } from 'src/components/images';

import { listApiKeys, hideNewApiKey } from 'src/actions/api-keys';
import { list as listSubaccounts } from 'src/actions/subaccounts';

import { Loading, Subaccount, TableCollection, ApiErrorBanner, ApiKeySuccessBanner, ShortKeyCode } from 'src/components';
import { filterBoxConfig } from './tableConfig';
import { selectKeysForAccount } from 'src/selectors/api-keys';
import { hasSubaccounts } from 'src/selectors/subaccounts';
import { setSubaccountQuery } from 'src/helpers/subaccounts';
import { LINKS } from 'src/constants';

const primaryAction = {
  content: 'Create API Key',
  Component: Link,
  to: '/account/api-keys/create'
};


export class ListPage extends Component {
  state = { copied: false };

  // only want to show the new key after a create
  UNSAFE_componentWillUnmount() {
    this.props.hideNewApiKey();
  }

  componentDidMount() {
    this.props.listApiKeys();
    if (hasSubaccounts && this.props.subaccounts.length === 0) {
      this.props.listSubaccounts();
    }
  }

  getLabel = ({ canCurrentUserEdit, id, subaccount_id, label }) => {
    if (canCurrentUserEdit) {
      return <Link to={`/account/api-keys/edit/${id}${setSubaccountQuery(subaccount_id)}`}>{label}</Link>;
    } else {
      return <Link to={`/account/api-keys/view/${id}${setSubaccountQuery(subaccount_id)}`}>{label}</Link>;
    }
  }

  getRowData = (key) => {
    const { short_key, subaccount_id, subaccount_name } = key;
    const { hasSubaccounts } = this.props;
    const rowData = [
      this.getLabel(key),
      <ShortKeyCode shortKey={short_key} />
    ];

    if (hasSubaccounts) {
      rowData.push(<Subaccount id={subaccount_id} name={subaccount_name}/>);
    }

    return rowData;
  };

  getColumns = () => {
    const { hasSubaccounts } = this.props;
    const columns = [
      { label: 'Name', width: '40%', sortKey: 'label' },
      { label: 'Key' }
    ];

    if (hasSubaccounts) {
      columns.push({ label: 'Subaccount', width: '20%', sortKey: 'subaccount_id' });
    }

    return columns;
  }

  onReloadApiBanner = () => {
    this.props.listApiKeys();
  };

  renderBanner() {
    return (
      <ApiKeySuccessBanner
        title="New API Key"
      />
    );
  }

  renderCollection() {
    const { keys } = this.props;

    return (
      <TableCollection
        columns={this.getColumns()}
        getRowData={this.getRowData}
        pagination={true}
        rows={keys}
        filterBox={filterBoxConfig}
        defaultSortColumn='label'
        defaultSortDirection='desc'
      />
    );
  }

  renderError() {
    return (
      <ApiErrorBanner
        errorDetails={this.props.error.message}
        message="Sorry, we seem to have had some trouble loading your API keys."
        reload={this.onReloadApiBanner}
      />
    );
  }

  render() {
    const { error, loading, newKey, keys } = this.props;

    if (loading) {
      return <Loading />;
    }

    return (
      <Page primaryAction={primaryAction} title='API Keys' empty={{
        show: keys.length === 0,
        image: Setup,
        content: <p>Create an API key you can use to access our REST or SMTP API services.</p>,
        secondaryAction: {
          content: 'View our API Docs',
          external: true,
          to: LINKS.API_DOCS
        }}}>
        {newKey && this.renderBanner()}
        {error ? this.renderError() : this.renderCollection()}
      </Page>
    );
  }
}

const mapStateToProps = (state) => {
  const { error, newKey, keysLoading } = state.apiKeys;
  return {
    hasSubaccounts: hasSubaccounts(state),
    subaccounts: state.subaccounts.list,
    keys: selectKeysForAccount(state),
    error,
    newKey,
    loading: keysLoading
  };
};

export default connect(mapStateToProps, { listApiKeys, listSubaccounts, hideNewApiKey })(ListPage);
