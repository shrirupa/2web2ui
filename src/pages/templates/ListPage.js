import React from 'react';
import { Link } from 'react-router-dom';
import { SubaccountTag, TableCollection, ApiErrorBanner, Loading } from 'src/components';
import { Templates } from 'src/components/images';
import { Page } from '@sparkpost/matchbox';
import { Name, Status, Actions, LastUpdated } from './components/ListComponents';
import { resolveTemplateStatus } from 'src/helpers/templates';

import useSparkpostApiRequest from 'src/actions/helpers/useSparkpostApiRequest';

function ListPage(props) {
  const { canModify, hasSubaccounts } = props;

  const { pending, results = [], retry, error } = useSparkpostApiRequest({
    method: 'GET',
    url: '/v1/templates'
  });

  if (pending) {
    return <Loading />;
  }

  const primaryAction = canModify ? {
    content: 'Create Template',
    to: '/templates/create',
    Component: Link
  } : null;

  return (
    <Page
      primaryAction={primaryAction}
      title='Templates'
      empty={{
        show: !error && results.length === 0,
        image: Templates,
        title: 'Manage your email templates',
        content: <p>Build, test, preview and send your transmissions.</p>
      }} >
      {error
        ? <ErrorBanner error={error} retry={retry} />
        : <Collection templates={results} hasSubaccounts={hasSubaccounts} />
      }
    </Page>
  );
}

function Collection({ templates, hasSubaccounts }) {
  const columns = [
    { label: 'Name', width: '28%', sortKey: 'name' },
    { label: 'Status', width: '18%', sortKey: (template) => [resolveTemplateStatus(template).publishedWithChanges, template.published]},
    ...(hasSubaccounts ? [{
      label: 'Subaccount', width: '18%', sortKey: (template) => [template.subaccount_id, template.shared_with_subaccounts]
    }] : []),
    { label: 'Last Updated', sortKey: 'last_update_time' },
    null
  ];

  const getRowData = ({ shared_with_subaccounts, ...rowData }) => {
    const { subaccount_id } = rowData;

    const subaccountCell = subaccount_id || shared_with_subaccounts
      ? <SubaccountTag all={shared_with_subaccounts} id={subaccount_id} />
      : null;

    return [
      <Name {...rowData} />,
      <Status {...rowData} />,
      ...(hasSubaccounts ? [subaccountCell] : []),
      <LastUpdated {...rowData}/>,
      <Actions {...rowData} />
    ];
  };

  return (
    <TableCollection
      columns={columns}
      rows={templates}
      getRowData={getRowData}
      pagination
      filterBox={{
        show: true,
        exampleModifiers: ['id', 'name'],
        itemToStringKeys: ['name', 'id', 'subaccount_id']
      }}
      defaultSortColumn='name'
    />
  );
}

function ErrorBanner({ error, retry }) {
  return (
    <ApiErrorBanner
      message={'Sorry, we seem to have had some trouble loading your templates.'}
      errorDetails={error.message}
      reload={retry}
    />
  );
}

export default ListPage;
