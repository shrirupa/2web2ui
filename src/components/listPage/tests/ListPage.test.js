import React from 'react';
import { shallow } from 'enzyme';

import ListPage from '../ListPage';

import * as fixtures from './fixtures';

describe('ListPage', () => {
  let baseProps;
  let onDelete;
  let onEdit;
  let items;
  let error;

  const subject = (props) => shallow(<ListPage {...baseProps} {...props} />);

  const renderRowWithActions = (row) => ({
    fields: baseProps.renderRow(row),
    actions: {
      editRoute: '/sprocket/1',
      deletable: true,
      customActions: [{ content: 'Flocculate', to: '/sprocket/1/flocculate' }]
    }
  });

  beforeEach(() => {
    baseProps = {
      columns: fixtures.columns,
      renderRow: ({ name, diam, scale }) => [name, diam, scale],
      loadItems: jest.fn(),
      noun: fixtures.noun,
      primaryAction: { to: '/sprocket/new' }
    };
    onEdit = jest.fn();
    onDelete = jest.fn().mockResolvedValue({});
    items = [...fixtures.items];
    error = { ...fixtures.error };
  });

  it('should render while loading', () => {
    expect(subject({ loading: true }).find('Loading')).toExist();
  });

  it('should render an error', () => {
    expect(subject({ error }).find('ApiErrorBanner')).toExist();
  });

  it('should render items', () => {
    expect(subject({ items }).find('TableCollection')).toExist();
  });

  it('should render a filter box', () => {
    expect(
      subject({ items, filterBox: fixtures.filterBox })
        .find('TableCollection')
        .prop('filterBox')
    ).toMatchObject({
      show: true,
      ...fixtures.filterBox
    });
  });

  it('should append item actions', () => {
    const wrapper = subject({
      items,
      onEdit,
      onDelete,
      renderRow: renderRowWithActions
    });
    const renderRowFn = wrapper.find('TableCollection').prop('getRowData');
    const row = renderRowFn(items[0]);
    expect(row).toHaveLength(Object.keys(items[0]).length + 1);
    expect(row).toMatchSnapshot();
  });

  it('should show the delete modal on request', () => {
    const wrapper = subject({ items, onEdit, onDelete });
    wrapper.instance().showDeleteModal(items[0]);
    expect(wrapper.find('DeleteModal').prop('open')).toBeTruthy();
  });

  it('should format a data-dependant custom delete message', () => {
    const renderDeleteWarning = (item) => (
      <p>{`${item.name} will no longer be able to sing!`}</p>
    );
    const wrapper = subject({ items, onDelete, renderDeleteWarning });
    wrapper.instance().showDeleteModal(items[0]);
    expect(wrapper.find('DeleteModal').prop('content')).toEqual(
      renderDeleteWarning(items[0])
    );
  });

  it('should call onDelete on confirm', () => {
    const wrapper = subject({ items, onEdit, onDelete });
    wrapper.find('DeleteModal').prop('onDelete')();
    expect(onDelete).toHaveBeenCalledTimes(1);
  });

  it('should dismiss the delete modal on cancel', () => {
    const wrapper = subject({ items, onEdit, onDelete });
    wrapper.instance().showDeleteModal(items[0]);
    wrapper.find('DeleteModal').prop('onCancel')();
    expect(onDelete).not.toHaveBeenCalled();
    expect(wrapper.find('DeleteModal').prop('open')).toBeFalsy();
  });

  it('should render custom primary action content', () => {
    const content = 'Make a thang';
    expect(subject({ primaryAction: { content, ...baseProps.primaryAction }}).find('Page').prop('primaryAction').content).toEqual(content);
  });
});
