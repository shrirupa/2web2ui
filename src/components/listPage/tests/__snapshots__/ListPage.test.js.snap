// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPage should append item actions 1`] = `
Array [
  "m3 bolt",
  3,
  "metric",
  <Actions
    customActions={
      Array [
        Object {
          "content": "Flocculate",
          "to": "/sprocket/1/flocculate",
        },
      ]
    }
    deletable={true}
    editRoute="/sprocket/1"
    item={
      Object {
        "diam": 3,
        "name": "m3 bolt",
        "scale": "metric",
      }
    }
    onDelete={[Function]}
  />,
]
`;
